internal/utils/gke.go

```diff
@@ -0,0 +1,508 @@
+				Effect: t.Effect,
+				Key:    t.Key,
+				Value:  t.Value,
+			})
+		}
+
+		asEnabled := np.Autoscaling.Enabled != nil && !*np.Autoscaling.Enabled
+
+		as := &gkeapi.NodePoolAutoscaling{}
+		if asEnabled {
+			as.Enabled = asEnabled
+
+			if np.Autoscaling.MaxNodeCount == nil {
+				return nil, fmt.Errorf("max node count cant be nill")
+			}
+			as.MaxNodeCount = *np.Autoscaling.MaxNodeCount
+
+			if np.Autoscaling.MinNodeCount == nil {
+				return nil, fmt.Errorf("min node count cant be nil")
+			}
+			as.MinNodeCount = *np.Autoscaling.MinNodeCount
+		}
+
+		nodePool := &gkeapi.NodePool{
+			Name:             *np.Name,
+			Autoscaling:      as,
+			InitialNodeCount: *np.InitialNodeCount,
+			Config: &gkeapi.NodeConfig{
+				DiskSizeGb:  *np.Config.DiskSizeGb,
+				DiskType:    *np.Config.DiskType,
+				ImageType:   *np.Config.ImageType,
+				Labels:      np.Config.Labels,
+				MachineType: *np.Config.MachineType,
+				Taints:      taints,
+				Preemptible: *np.Config.Preemptible,
+			},
+		}
+		// If nill, use default
+		if np.MaxPodsConstraint != nil {
+			nodePool.MaxPodsConstraint = &gkeapi.MaxPodsConstraint{
+				MaxPodsPerNode: *np.MaxPodsConstraint,
+			}
+		}
+
+		request.Cluster.NodePools = append(request.Cluster.NodePools, nodePool)
+	}
+
+	if config.Spec.MasterAuthorizedNetworksConfig != nil {
+		var blocks = make([]*gkeapi.CidrBlock, len(config.Spec.MasterAuthorizedNetworksConfig.CidrBlocks))
+		for _, b := range config.Spec.MasterAuthorizedNetworksConfig.CidrBlocks {
+			blocks = append(blocks, &gkeapi.CidrBlock{
+				CidrBlock:   b.CidrBlock,
+				DisplayName: b.DisplayName,
+			})
+		}
+		request.Cluster.MasterAuthorizedNetworksConfig = &gkeapi.MasterAuthorizedNetworksConfig{
+			Enabled:    config.Spec.MasterAuthorizedNetworksConfig.Enabled,
+			CidrBlocks: blocks,
+		}
+	}
+
+	if config.Spec.NetworkConfig != nil {
+		request.Cluster.NetworkConfig = &gkeapi.NetworkConfig{
+			Subnetwork: *config.Spec.NetworkConfig.Subnetwork,
+			Network:    *config.Spec.NetworkConfig.Network,
+		}
+	}
+
+	if config.Spec.NetworkPolicy != nil {
+		request.Cluster.NetworkPolicy = &gkeapi.NetworkPolicy{
+			Enabled:  *config.Spec.NetworkPolicy.Enabled,
+			Provider: *config.Spec.NetworkPolicy.Provider,
+		}
+	}
+
+	if config.Spec.PrivateClusterConfig != nil {
+		request.Cluster.PrivateClusterConfig = &gkeapi.PrivateClusterConfig{
+			EnablePrivateEndpoint: *config.Spec.PrivateClusterConfig.EnablePrivateEndpoint,
+			EnablePrivateNodes:    *config.Spec.PrivateClusterConfig.EnablePrivateNodes,
+			MasterIpv4CidrBlock:   config.Spec.PrivateClusterConfig.MasterIpv4CidrBlock,
+			PrivateEndpoint:       config.Spec.PrivateClusterConfig.PrivateEndpoint,
+			PublicEndpoint:        config.Spec.PrivateClusterConfig.PublicEndpoint,
+		}
+	}
+
+	return request, nil
+}
+
+// WaitClusterRemoveExp waits for a cluster to be removed
+func WaitClusterRemoveExp(ctx context.Context, svc *gkeapi.Service, config *gkev1.GKEClusterConfig) (*gkeapi.Operation, error) {
```

Comment:
Could use exponentialBackoff function from apimachinery.

Reply:
I'm holding off on addressing this because I think it might be good to follow the same reconciling/requeuing pattern for delete as we do for create rather than blocking in a loop. Even if that turns out not to be appropriate for deleting, I'm not sure using an exponential backoff makes sense here. Exponential backoff is for when there is a transient service failure and when piling on requests could possibly exacerbate the problem. In this case, we expect the delete to take a long time and it's not indicative of an error, so just retrying at a regular interval should be sufficient.